File: src/main/java/com/youngjun/auth/core/api/security/filter/JwtAuthenticationProvider.java
package com.youngjun.auth.core.api.security.filter;

import com.youngjun.auth.core.domain.auth.Auth;
import com.youngjun.auth.core.domain.auth.AuthService;
import com.youngjun.auth.core.domain.token.TokenParser;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Component;

@Component
public class JwtAuthenticationProvider implements AuthenticationProvider {

    private final TokenParser tokenParser;

    private final AuthService authService;

    public JwtAuthenticationProvider(TokenParser tokenParser, AuthService authService) {
        this.tokenParser = tokenParser;
        this.authService = authService;
    }

    @Override
    public Authentication authenticate(Authentication authentication) {
        String token = (String) authentication.getPrincipal();
        try {
            String username = tokenParser.parseSubject(token);
            return createSuccessAuthentication(username);
        } catch (Exception ex) {
            throw new BadTokenException(ex.getMessage(), ex.getCause());
        }
    }

    private Authentication createSuccessAuthentication(String username) {
        Auth auth = authService.loadUserByUsername(username);
        return JwtAuthenticationToken.authenticated(auth);
    }

    @Override
    public boolean supports(Class<?> authentication) {
        return BearerTokenAuthenticationToken.class.isAssignableFrom(authentication);
    }

}

File: src/main/kotlin/com/youngjun/auth/core/api/security/JwtAuthenticationProvider.kt
package com.youngjun.auth.core.api.security

import com.youngjun.auth.core.api.application.UserService
import com.youngjun.auth.core.domain.token.TokenParser
import org.springframework.security.authentication.AccountExpiredException
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.authentication.DisabledException
import org.springframework.security.authentication.LockedException
import org.springframework.security.core.Authentication
import org.springframework.security.core.userdetails.UserDetails

class JwtAuthenticationProvider(
    private val tokenParser: TokenParser,
    private val userService: UserService,
) : AuthenticationProvider {
    override fun authenticate(authentication: Authentication): Authentication {
        val username = tokenParser.parseSubject(authentication.credentials as String)
        val user = userService.loadUserByUsername(username)
        check(user)
        return JwtAuthenticationToken.authenticated(user)
    }

    private fun check(user: UserDetails) {
        when {
            !user.isAccountNonLocked -> throw LockedException("User account is locked")
            !user.isEnabled -> throw DisabledException("User is disabled")
            !user.isAccountNonExpired -> throw AccountExpiredException("User account has expired")
        }
    }

    override fun supports(authentication: Class<*>): Boolean = BearerTokenAuthenticationToken::class.java.isAssignableFrom(authentication)
}
